pyro/infer/mcmc/nuts.py

```diff
@@ -150,24 +155,28 @@ def _build_basetree(self, z, r, z_grads, log_slice, direction, energy_current):
             z, r, self._potential_energy, self._inverse_mass_matrix, step_size, z_grads=z_grads)
         r_new_flat = torch.cat([r_new[site_name].reshape(-1) for site_name in sorted(r_new)])
         energy_new = potential_energy + self._kinetic_energy(r_new)
-        sliced_energy = energy_new + log_slice
 
-        # As a part of the slice sampling process (see below), along the trajectory
-        #     we eliminate states which p(z, r) < u, or dE > 0.
-        # Due to this elimination (and stop doubling conditions),
-        #     the size of binary tree might not equal to 2^tree_depth.
-        tree_size = 1 if sliced_energy <= 0 else 0
         # Special case: Set diverging to True and accept prob to 0 if the
         # diverging trajectory returns `NaN` energy (e.g. in the case of
         # evaluating log prob of a value simulated using a large step size
         # for a constrained sample site).
-        if torch_isnan(energy_new):
-            diverging = True
-            accept_prob = energy_new.new_tensor(0.0)
+        # This is obtained by setting energy_new to +inf when it is nan.
+        energy_new = energy_new.new_tensor(float("inf")) if torch_isnan(energy_new) else energy_new
+        sliced_energy = energy_new + log_slice
+        diverging = (sliced_energy > self._max_sliced_energy)
+        delta_energy = energy_new - energy_current
+        accept_prob = (-delta_energy).exp().clamp(max=1.0)
+
+        if self.use_multinomial_sampling:
+            tree_size = -sliced_energy
```

Comment:
This variable name is unintuitive for the multinomial sampling case - we should consider renaming this and adding a comment as to what it corresponds to in the multinomial and slice sampling case.

Reply:
I have added a comment at the definition of TreeInfo. Could you suggest me another name for it? :)

Reply:
How about using `weight` - that might convey the desired use case?

Reply:
That is a good name!

Reply:
Thanks @neerajprad !